pkg/drivers/sqlserver/sqlserver.go

```diff
@@ -0,0 +1,282 @@
+package sqlserver
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	mssql "github.com/denisenkom/go-mssqldb"
+	"github.com/rancher/kine/pkg/drivers/generic"
+	"github.com/rancher/kine/pkg/logstructured"
+	"github.com/rancher/kine/pkg/logstructured/sqllog"
+	"github.com/rancher/kine/pkg/server"
+	"github.com/rancher/kine/pkg/tls"
+)
+
+const (
+	defaultDSN = "sqlserver://sa:"
+)
+
+var (
+	schema = []string{
+		`if not exists (SELECT * FROM INFORMATION_SCHEMA.TABLES
+           WHERE TABLE_NAME = N'kine')
+		begin
+ 			create table kine (
+				id int primary key identity (1, 1),
+				name varchar(630),
+				created int,
+				deleted int,
+				create_revision int,
+				prev_revision int,
+				lease int,
+				value varbinary(max),
+				old_value varbinary(max) )
+		end 
+		`,
+		`if not exists ( select * from sys.indexes
+			where name = 'kine_name_index' and
+			object_id = OBJECT_ID('kine')) 
+		begin
+    		create nonclustered index kine_name_index on kine (name)
+		end
+		`,
+		`if not exists (
+			select *
+			from sys.indexes
+			where name = 'kine_name_prev_revision_uindex' and
+			object_id = OBJECT_ID('kine')
+		) begin
+		create unique index kine_name_prev_revision_uindex on kine (name, prev_revision)
+		end
+		`,
+	}
+	createDB = "create database "
+	columns  = "kv.id as theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value"
+	revSQL   = `
+		SELECT TOP 1 rkv.id
+		FROM kine rkv
+		ORDER BY rkv.id
+		DESC`
+	compactRevSQL = `
+		SELECT TOP 1 crkv.prev_revision
+		FROM kine crkv
+		WHERE crkv.name = 'compact_rev_key'
+		ORDER BY crkv.id DESC`
+
+	idOfKey = `
+		AND mkv.id <= ? AND mkv.id > (
+			SELECT TOP 1 ikv.id
+			FROM kine ikv
+			WHERE
+				ikv.name = ? AND
+				ikv.id <= ?
+			ORDER BY ikv.id DESC )`
+
+	listSQL = fmt.Sprintf(`SELECT TOP 100 PERCENT (%s)[a], (%s)[b], %s
```

Comment:
I"m confused why there is a `TOP 100 PERCENT` in here by default - is that just so you can hack in a limit by with string replacement on the outer SELECT TOP statement without worrying affecting the inner SELECT?

As I read it, you should be safe to do string replacement on a plain SELECT since you're limiting the number of replacements to 1 which should only catch the outer SELECT? I admit the nested SQL printfs are a bit hard to read so I might be missing something.

Reply:
When the server boots up it tries to range over keys, and that translates itself as the following query

```
TRAC[0004] QUERY [/registry/mutatingwebhookconfigurations/% 3 false] : SELECT ( SELECT TOP 1 rkv.id FROM kine rkv ORDER BY rkv.id DESC)[a], ( SELECT TOP 1 crkv.prev_revision FROM kine crkv WHERE crkv.name = 'compact_rev_key' ORDER BY crkv.id DESC)[b], kv.id as theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value FROM kine kv JOIN ( SELECT MAX(mkv.id) as id FROM kine mkv WHERE mkv.name LIKE @p1 AND mkv.id <= @p2 GROUP BY mkv.name) maxkv ON maxkv.id = kv.id WHERE ( kv.deleted = 0 OR 'true' = @p3 ) ORDER BY kv.id ASC

ERRO[0004] error while range on /registry/mutatingwebhookconfigurations /registry/mutatingwebhookconfigurations: mssql: The ORDER BY clause is invalid in views, inline functions, derived tables, subqueries, and common table expressions, unless TOP, OFFSET or FOR XML is also specified.
```

sqlserver syntax forces me to add the extra arguments to ensure I can get all the keys back.

May be it is just my limited knowledge of sqlserver.

Because of this the ApplyLimit function needs to handle the extra `TOP 100 PERCENT`, i.e., replace it and apply the correct limit for the specific call.